File: packages/authentication/go/cookie.go
package authentication

import (
	"context"
	"net/url"
	"strings"

	botiva "github.com/aimtune/botiva/core"
)

// CookieAuthenticator extracts a credential from a named cookie in the request
// headers, then delegates verification to an inner authenticator. Browsers
// attach cookies automatically, so this needs no client-side token plumbing —
// pair it with an HMAC-JWT or static-token verifier.
//
//	authentication.NewCookieAuthenticator("botiva_session", jwtAuth)
type CookieAuthenticator struct {
	cookie string
	inner  botiva.Authenticator
}

// NewCookieAuthenticator panics on an empty cookie name or nil inner.
func NewCookieAuthenticator(cookie string, inner botiva.Authenticator) *CookieAuthenticator {
	if cookie == "" {
		panic("authentication: CookieAuthenticator requires a cookie name")
	}
	if inner == nil {
		panic("authentication: CookieAuthenticator requires an inner authenticator")
	}
	return &CookieAuthenticator{cookie: cookie, inner: inner}
}

// Authenticate reads the cookie (falling back to any existing ac.Token) and
// forwards to the inner authenticator.
func (a *CookieAuthenticator) Authenticate(ctx context.Context, ac botiva.AuthContext) (botiva.AuthResult, error) {
	if token, ok := ParseCookies(ac.Headers["cookie"])[a.cookie]; ok {
		ac.Token = token
	}
	return a.inner.Authenticate(ctx, ac)
}

// ParseCookies parses a Cookie header value into a name → value map.
func ParseCookies(header string) map[string]string {
	out := map[string]string{}
	if header == "" {
		return out
	}
	for _, pair := range strings.Split(header, ";") {
		eq := strings.IndexByte(pair, '=')
		if eq < 0 {
			continue
		}
		name := strings.TrimSpace(pair[:eq])
		if name == "" {
			continue
		}
		value := strings.TrimSpace(pair[eq+1:])
		if decoded, err := url.QueryUnescape(value); err == nil {
			value = decoded
		}
		out[name] = value
	}
	return out
}

File: packages/authentication/ts/src/cookie.ts
import type { AuthContext, AuthResult, Authenticator } from "@botiva/core";

export interface CookieAuthenticatorOptions {
    /** Cookie name carrying the credential (e.g. "botiva_session"). */
    cookie: string;
    /** Verifier the extracted cookie value is handed to (as `ctx.token`). */
    inner: Authenticator;
}

/**
 * Extracts a credential from a named cookie in the request headers, then
 * delegates verification to an inner authenticator. Browsers attach cookies
 * automatically, so this needs no client-side token plumbing — pair it with
 * `HmacJwtAuthenticator` or `StaticTokenAuthenticator`:
 *
 *   new CookieAuthenticator({
 *       cookie: "botiva_session",
 *       inner: new HmacJwtAuthenticator({ secret }),
 *   })
 */
export class CookieAuthenticator implements Authenticator {
    readonly #cookie: string;
    readonly #inner: Authenticator;

    constructor(opts: CookieAuthenticatorOptions) {
        if (!opts?.cookie) throw new Error("CookieAuthenticator requires a cookie name.");
        if (!opts?.inner) throw new Error("CookieAuthenticator requires an inner authenticator.");
        this.#cookie = opts.cookie;
        this.#inner = opts.inner;
    }

    authenticate(ctx: AuthContext): AuthResult | Promise<AuthResult> {
        const token = parseCookies(ctx.headers?.cookie)[this.#cookie];
        // Keep any existing token as a fallback (e.g. query/hello/Bearer).
        return this.#inner.authenticate({ ...ctx, token: token ?? ctx.token });
    }
}

/** Parse a `Cookie:` header value into a name → value map. */
export function parseCookies(header: string | undefined): Record<string, string> {
    const out: Record<string, string> = {};
    if (!header) return out;
    for (const pair of header.split(";")) {
        const eq = pair.indexOf("=");
        if (eq < 0) continue;
        const name = pair.slice(0, eq).trim();
        if (!name) continue;
        const value = pair.slice(eq + 1).trim();
        try {
            out[name] = decodeURIComponent(value);
        } catch {
            out[name] = value;
        }
    }
    return out;
}
